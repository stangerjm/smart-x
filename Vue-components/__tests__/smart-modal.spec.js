import {mount} from 'vue-test-utils'
import smartModal from '../components/smart-modal'
import {shallow} from 'vue-test-utils'
import {createRenderer} from 'vue-server-renderer'

function mountModal(props) {
  if (!props) {
    props = {
      propsData: {
        modalTitle: 'Test Modal'
      }
    }
  }

  return mount(smartModal, props);
}

describe('smart-modal.vue', () => {
  it('has "smart-modal" as a class value', () => {
    const wrapper = mountModal();

    expect(wrapper.vm.$el.className).toEqual('smart-modal');
  });

  it('requries a string to be passed into a "modalTitle" property', () => {
    const wrapper = mountModal();
    let modalTitleProp = wrapper.vm.$options.props.modalTitle;

    expect(modalTitleProp.type).toEqual(String);
    expect(modalTitleProp.required).toBeTruthy();
  });

  it('allows content to be passed into a "main" and "footer" slot', () => {
    let msg = 'TEST';
    let mainContent = '<h1>' + msg + '</h1>';
    let footerContent = '<span>' + msg + '</span>';

    const wrapper = mount(smartModal, {
      propsData: {
        modalTitle: 'Test'
      },
      slots: {
        main: mainContent,
        footer: footerContent
      }
    });

    let main = wrapper.vm.$el.querySelector('.smart-modal--content > .smart-modal--body');
    let footer = wrapper.vm.$el.querySelector('.smart-modal--content > .smart-modal--footer');

    expect(main.innerHTML).toEqual(mainContent);
    expect(footer.innerHTML).toEqual(footerContent);
  });

  it('will close if the "smart-modal--close" button is clicked, or if the escape button is hit', () => {
    const wrapper = mountModal();

    //simulate pressing escape button
    wrapper.trigger('keydown', {
      keyCode: 27
    });

    expect(wrapper.vm.$el.style.display).toEqual('none');

    wrapper.vm.open();

    expect(wrapper.vm.$el.style.display).toEqual('');

    //simulate clicking the button
    wrapper.vm.close();

    expect(wrapper.vm.$el.style.display).toEqual('none');
  });

  it('has the same HTML structure', () => {
    const renderer = createRenderer();
    const wrapper = shallow(smartModal, {
      propsData: {
        modalTitle: 'Test'
      }
    });

    renderer.renderToString(wrapper.vm, (err, str) => {
      if (err) {
        throw new Error(err);
      }
      expect(str).toMatchSnapshot();
    });
  });
});