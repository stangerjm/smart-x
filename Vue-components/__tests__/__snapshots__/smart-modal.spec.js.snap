// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`smart-modal.vue has the same HTML structure 1`] = `
<section role="dialog" aria-labelledby="be-a-trooper" aria-describedby="Join the Washington State Patrol" class="smart-modal">
  <div class="smart-modal--content">
    <header class="smart-modal--head">
      <button class="smart-modal--close">CLOSE</button>
      <h2 class="smart-modal--title">Test</h2>
    </header>
    <main class="smart-modal--body"></main>
    <footer class="smart-modal--footer"></footer>
  </div>
</section>
`;
